clash-ghc/src-ghc/Clash/GHC/PartialEval/Eval.hs

```diff
@@ -0,0 +1,603 @@
+etaExpand term = do
+  tcm <- getTyConMap
+
+  case collectArgs term of
+    x@(Data dc, _) -> expand tcm (dcType dc) x
+    x@(Prim pr, _) -> expand tcm (primType pr) x
+    _ -> pure term
+ where
+  etaNameOf =
+    either (pure . Right) (fmap Left . getUniqueId "eta")
+
+  expand tcm ty (tm, args) = do
+    let (missingTys, _) = splitFunForallTy (applyTypeToArgs tm tcm ty args)
+    missingArgs <- traverse etaNameOf missingTys
+
+    pure $ mkAbstraction
+      (mkApps term (fmap (bimap Var VarTy) missingArgs))
+      missingArgs
+
+evalLam :: Id -> Term -> Eval Value
+evalLam i x = do
+  varTy <- evalType (varType i)
+  let i' = i { varType = varTy }
+  env <- getLocalEnv
+
+  pure (VLam i' x env)
+
+evalTyLam :: TyVar -> Term -> Eval Value
+evalTyLam i x = do
+  varTy <- evalType (varType i)
+  let i' = i { varType = varTy }
+  env <- getLocalEnv
+
+  pure (VTyLam i' x env)
+
+evalApp :: Term -> Arg Term -> Eval Value
+evalApp x y
+  | Data dc <- f
+  = if fullyApplied (dcType dc) args
+      then do
+        argThunks <- delayArgs args
+        VData dc argThunks <$> getLocalEnv
+
+      else etaExpand term >>= eval
+
+  | Prim pr <- f
+  , prArgs  <- fst $ splitFunForallTy (primType pr)
+  , numArgs <- length prArgs
+  = case compare (length args) numArgs of
+      LT ->
+        etaExpand term >>= eval
+
+      EQ -> do
+        argThunks <- delayArgs args
+        let tyVars = lefts prArgs
+            tyArgs = rights args
+
+        withTyVars (zip tyVars tyArgs) (evalPrimOp pr argThunks)
+
+      GT -> do
+        let (pArgs, rArgs) = splitAt numArgs args
+        pArgThunks <- delayArgs pArgs
+        primRes <- evalPrimOp pr pArgThunks
+        rArgThunks <- delayArgs rArgs
+
+        foldM applyArg primRes rArgThunks
+
+  | otherwise
+  = preserveFuel $ do
+      evalF <- eval f
+      argThunks <- delayArgs args
+      foldM applyArg evalF argThunks
+ where
+  term = either (App x) (TyApp x) y
+  (f, args, _ticks) = collectArgsTicks term
+
+evalLetrec :: [LetBinding] -> Term -> Eval Value
+evalLetrec bs x = do
+  -- Determine if a binding should be kept in a letrec or inlined. We keep
+  -- bindings which perform work to prevent duplication of registers etc.
+  (keep, inline) <- foldM evalScc ([], []) (Util.sccLetBindings bs)
+  eX <- withIds (keep <> inline) (eval x)
+
+  case keep of
+    [] -> pure eX
+    _  -> pure (VNeutral (NeLetrec keep eX))
+ where
+  evalBind (i, y) = do
+    iTy <- evalType (varType i)
+    eY <- delayEval y
+
+    pure (i { varType = iTy }, eY)
+
+  evalScc (k, i) = \case
+    AcyclicSCC y -> do
+      eY <- evalBind y
+      workFree <- workFreeValue (snd eY)
+
+      if workFree then pure (k, eY:i) else pure (eY:k, i)
+
+    CyclicSCC ys -> do
+      eYs <- traverse evalBind ys
+      pure (eYs <> k, i)
+
+evalCase :: Term -> Type -> [Alt] -> Eval Value
+evalCase term ty as
+  -- There is only one option, the subject need not be evaluated.
+  | [(DefaultPat, tm)] <- as
+  = eval tm
```

Comment:
What if the subject is `_|_`?

Reply:
To clarify; I'm not worried that this is incorrect, but we might end up generating larger designs this way.

Reply:
Yes, reasonable point. The new evaluator currently considers `_|_` but only after checking if there is only `_` as a pattern. It would probably better to check subject first, because there might be a long chain of `case ... of _ -> ...` in core which can be completely skipped if the first subject is `_|_`.

Reply:
Okay, changed to check for `_|_` first in case expressions. It also occurred to me that there isn't really a reason to check for just `DefaultPat` first (since that was just to avoid evaluating the subject) so I removed that special path and tidied it up a little.